File: API/src/main/java/com/aklaa/api/services/contract/UserService.java
package com.aklaa.api.services.contract;

import com.aklaa.api.dtos.response.UserDTO;
import com.aklaa.api.dtos.response.UserListResponseDTO;
import com.aklaa.api.model.enums.UserType;
import org.springframework.data.domain.Pageable;

/**
 * Service interface for user management operations.
 * <p>
 * This service provides functionality for retrieving, filtering, and updating users,
 * as well as refreshing authentication contexts.
 * </p>
 */
public interface UserService {
    
    /**
     * Retrieves and filters users based on search criteria, user type, and pagination.
     * <p>
     * This method searches across user first names, last names, and email addresses.
     * It filters by user type if provided and returns paginated results.
     * </p>
     *
     * @param search the search term to filter users (searches in firstName, lastName, and email)
     * @param type the user type name to filter by (e.g., "ADMIN", "USER")
     * @param pageable the pagination information including page number, size, and sorting
     * @return a {@link UserListResponseDTO} containing the filtered users, total elements, and total pages
     */
    UserListResponseDTO getUsers(String search, String type, Pageable pageable);
    
    /**
     * Retrieves a specific user by their ID.
     * <p>
     * This method fetches the user details from the database.
     * </p>
     *
     * @param id the ID of the user to retrieve
     * @return a {@link UserDTO} containing the user's data
     * @throws NoSuchElementException if the user is not found
     */
    UserDTO get(Long id);
    
    /**
     * Updates a user's role/type.
     * <p>
     * This method updates the user's type in the database and refreshes the
     * security authentication context to reflect the new role immediately.
     * </p>
     *
     * @param id the ID of the user whose role should be updated
     * @param userType the new user type to assign
     * @return a {@link UserDTO} containing the updated user's data
     * @throws NoSuchElementException if the user is not found
     */
    UserDTO updateUserRole(Long id, UserType userType);
    
    /**
     * Refreshes the authentication context for a user.
     * <p>
     * This method reloads the user details from the database and updates
     * the Spring Security context with the fresh user information and authorities.
     * This is typically called after updating user roles to ensure the changes
     * are reflected in the current security context.
     * </p>
     *
     * @param email the email address of the user whose authentication should be refreshed
     */
    void refreshAuthentication(String email);
}

File: Web/src/services/userService.js
import api from '../api/axiosConfig';

/**
 * User Service
 * Handles all user-related API calls
 */

const userService = {
  /**
   * Get all users (admin)
   * @param {Object} params - Query parameters
   * @returns {Promise} Users list
   */
  getUsers: async (params = {}) => {
    const response = await api.get('/users', { params });
    return response;
  },

  /**
   * Update user email
   * @param {Object} emailData - New email data
   * @returns {Promise} Update response
   */
  updateEmail: async (emailData) => {
    return api.put('/users/email', emailData);
  },

  /**
   * Update user role (admin)
   * @param {number|string} userId - User ID
   * @param {string} newRole - New role
   * @returns {Promise} Update response
   */
  updateUserRole: async (userId, newRole) => {
    return api.put(`/users/${userId}`, null, { params: { type: newRole } });
  },

  /**
   * Enable user account (admin)
   * @param {number|string} userId - User ID
   * @returns {Promise} Enable response
   */
  enableUser: async (userId) => {
    return api.put(`/users/enable/${userId}`);
  },

  /**
   * Delete user (admin or self)
   * @param {number|string} userId - User ID (optional, deletes current user if not provided)
   * @returns {Promise} Delete response
   */
  deleteUser: async (userId = null) => {
    if (userId) {
      return api.delete(`/users/${userId}`);
    }
    return api.delete('/users');
  },

  /**
   * Export user data
   * @returns {Promise} Exported data blob
   */
  exportUserData: async () => {
    const response = await api.get('/users/data/export', { responseType: 'blob' });
    return response;
  },
};

export default userService;
